composer/utils/object_store/gs_object_store.py

```diff
@@ -0,0 +1,211 @@
+            dest (str, optional): The destination path in the cloud storage bucket where the file will be saved.
+                If not provided or an empty string is given, the file will be uploaded to the root of the bucket with the same
+                name as the source file. Default is an empty string.
+            callback: optional
+        """
+
+        dest = str(src) if dest == '' else dest
+        blob = self.bucket.blob(self.get_key(dest))
+        blob.upload_from_filename(src)
+
+    def download_object(
+        self,
+        src: str,
+        dest: Union[str, pathlib.Path],
+        overwrite: bool = False,
+        callback: Optional[Callable[[int, int], None]] = None,
+    ):
+        """Downloads an object from the specified source in the cloud storage bucket and saves it to the given destination.
+
+        Args:
+            src (str): The path to the object in the cloud storage bucket that needs to be downloaded.
+            dest (Union[str, pathlib.Path]): The destination path where the object will be saved locally. It can be a
+                string representing the file path or a pathlib.Path object.
+            overwrite (bool, optional): If set to True, the function will overwrite the destination file if it already
+                exists. If set to False, and the destination file exists, a FileExistsError will be raised. Default is False.
+            callback (Callable[[int, int], None], optional): A callback function that can be used to track the progress of
+                the download. It takes two integer arguments - the number of bytes downloaded and the total size of the
+                object. Default is None.
+
+        Raises:
+            FileExistsError: If the destination file already exists and the `overwrite` parameter is set to False.
+        """
+        if os.path.exists(dest) and not overwrite:
+            raise FileExistsError(f'The file at {dest} already exists and overwrite is set to False.')
+
+        dirname = os.path.dirname(dest)
+        if dirname:
+            os.makedirs(dirname, exist_ok=True)
+        tmp_path = str(dest) + f'.{uuid.uuid4()}.tmp'
+
+        try:
+            try:
+                blob = self.bucket.blob(self.get_key(src))
+                blob.download_to_filename(tmp_path)
+            except Exception as e:
+                _reraise_gs_errors(self.get_uri(src), e)
+        except:
+            # Make a best effort attempt to clean up the temporary file
+            try:
+                os.remove(tmp_path)
+            except OSError:
+                pass
+            raise
+        else:
+            if overwrite:
+                os.replace(tmp_path, dest)
+            else:
+                os.rename(tmp_path, dest)
+
+    def list_objects(self, prefix: Optional[str] = None) -> List[str]:
+
+        if prefix is None:
+            prefix = ''
+        prefix = self.get_key(prefix)
+
+        object_names = []
```

Comment:
this variable is not actually object names right? its objects? and then return gets the names

Reply:
Done